flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -0,0 +1,207 @@
+    return lbounds;
+  }
+
+  bool lboundsAllOne() const { return lbounds.empty(); }
+
+  llvm::SmallVector<mlir::Value, 4> extents;
+  llvm::SmallVector<mlir::Value, 4> lbounds;
+};
+
+/// Expressions with rank > 0 have extents. They may also have lbounds that are
+/// not 1.
+struct ArrayBoxValue : public AbstractBox, public AbstractArrayBox {
+  ArrayBoxValue(mlir::Value addr, llvm::ArrayRef<mlir::Value> extents,
+                llvm::ArrayRef<mlir::Value> lbounds = {})
+      : AbstractBox{addr}, AbstractArrayBox{extents, lbounds} {}
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const ArrayBoxValue &);
+  void dump() const { operator<<(llvm::errs(), *this); }
+};
+
+/// Expressions of type CHARACTER and with rank > 0.
+struct CharArrayBoxValue : public CharBoxValue, public AbstractArrayBox {
+  CharArrayBoxValue(mlir::Value addr, mlir::Value len,
+                    llvm::ArrayRef<mlir::Value> extents,
+                    llvm::ArrayRef<mlir::Value> lbounds = {})
+      : CharBoxValue{addr, len}, AbstractArrayBox{extents, lbounds} {}
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const CharArrayBoxValue &);
+  void dump() const { operator<<(llvm::errs(), *this); }
+};
+
+/// Expressions that are procedure POINTERs may need a set of references to
+/// variables in the host scope.
+struct ProcBoxValue : public AbstractBox {
+  ProcBoxValue(mlir::Value addr, mlir::Value context)
+      : AbstractBox{addr}, hostContext{context} {}
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const ProcBoxValue &);
+  void dump() const { operator<<(llvm::errs(), *this); }
+
+  mlir::Value hostContext;
+};
+
+/// In the generalized form, a boxed value can have a dynamic size, be an array
+/// with dynamic extents and lbounds, and take dynamic type parameters.
+struct BoxValue : public AbstractBox, public AbstractArrayBox {
+  BoxValue(mlir::Value addr) : AbstractBox{addr}, AbstractArrayBox{} {}
+  BoxValue(mlir::Value addr, mlir::Value len)
+      : AbstractBox{addr}, AbstractArrayBox{}, len{len} {}
+  BoxValue(mlir::Value addr, llvm::ArrayRef<mlir::Value> extents,
+           llvm::ArrayRef<mlir::Value> lbounds = {})
+      : AbstractBox{addr}, AbstractArrayBox{extents, lbounds} {}
+  BoxValue(mlir::Value addr, mlir::Value len,
+           llvm::ArrayRef<mlir::Value> params,
+           llvm::ArrayRef<mlir::Value> extents,
+           llvm::ArrayRef<mlir::Value> lbounds = {})
+      : AbstractBox{addr}, AbstractArrayBox{extents, lbounds}, len{len},
+        params{params.begin(), params.end()} {}
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &, const BoxValue &);
+  void dump() const { operator<<(llvm::errs(), *this); }
+
+  mlir::Value len;
+  llvm::SmallVector<mlir::Value, 2> params;
+};
+
+using RangeBoxValue = std::tuple<mlir::Value, mlir::Value, mlir::Value>;
```

Comment:
Fortran can just be ugly. In an array expression, we have a number of possible indexing options. The indices may be entirely elided (implied). Or we might have some indices of various forms: a scalar value, an array value, an elided marker, a triple "slice" operation.
The first three of these fit in the lower framework fairly cleanly. The triple is not as clean. Still, the interface to lower each component of the indices has to maintain the same call/return structure or else the expression converter will have to test for triple notation as the special case to handle distinctly everywhere it might legally appear in the traversal. I've chosen to standardize the traversal while specializing the values produced and consumed.

Reply:
> everywhere it might legally appear in the traversal

But does it appear elsewhere than in the ArrayRef and CoarrayRef  cases ? I see two places that checks for `getRange()` currently and I do not foresee others, especially outside of expression lowering.
Placing `RangeBoxValue` under `ExtentedValue` allows `RangeBoxValue` to be propagated in places where it does not make any sens (for instance, it cannot be returned by expression lowering to the bridge), which I think far out-numbers the places of the code where it makes sens to get such value.
For all other types unioned in `ExtentedValue`, it makes sens to ask about them: "What is there type ?", "What is the shape ?", "Is it in memory ?" but it does not make much sens for `RangeBoxValue`.

Anyway, if you see a need to propagate `RangeBoxValue` far away, then I am not against, I just do not see the use case right now and I would favor having the subscript lowering function returning a `std::variant<ExtentedValue, RangeBoxValue>` or similar.

Reply:
Note that an expression in an index position as an array argument is just like any other Fortran::evaluate::SomeExpr.

That said, I can refactor the recursive descent a bit and see if that clarifies or muddies the picture.